File: manual/hs/manual/app/Manual/Types/Unions.hs
{-# LANGUAGE DataKinds #-}

module Manual.Types.Unions (examples) where

import Foreign as F

import HsBindgen.Runtime.Union qualified as Union

import Manual.Tools
import Unions
import Unions.Unsafe

{-------------------------------------------------------------------------------
  Examples
-------------------------------------------------------------------------------}

examples :: IO ()
examples = do
    section "Unions"

    do
      let occupation = Union.set @"occupation_student" Student{
              student_university = nullPtr
            , student_year       = 2000
            }
      print $ Union.get @"occupation_student" occupation
      with occupation $ print_occupation 0

    do
      let occupation = Union.set @"occupation_employee" Employee{
              employee_company    = nullPtr
            , employee_supervisor = nullPtr
            , employee_salary     = 100_000
            }
      print $ Union.get @"occupation_employee" occupation
      with occupation $ print_occupation 1

File: manual/c/unions.c
#include "unions.h"
#include <stdio.h>

/* -------------------------------------------------------------------------- */
/* Unions */

void print_occupation(int tag, occupation* o) {
    switch(tag) {
        case 0:
            printf("student in year %d\n", o->student.year);
            break;
        case 1:
            printf("employee with salary %d\n", o->employee.salary);
            break;
        default:
            printf("unknown tag\n");
            break;
    };
}
